src/audio/pcm_converter/pcm_converter_hifi3.c

```diff
@@ -757,7 +807,523 @@ const struct pcm_func_map pcm_func_map[] = {
+			AE_SA32X2_IP(sample, outu, out);
+		}
+		AE_SA64POS_FP(outu, out);
+
+		/* process the left 1 sample to avoid memory access overrun */
+		if (n & 0x01) {
+			AE_LA24_IP(sample24, inu, in);
+			sample = AE_MOVINT32_FROMINT24X2(sample24);
+			AE_S32_L_IP(sample, (ae_int32 *)out,
+				    sizeof(ae_int32));
+		}
+
+		in = audio_stream_wrap(source, in);
+		out = audio_stream_wrap(sink, out);
+	}
+
+	return samples;
+}
+
+static int pcm_convert_s24_c32_to_s24_c24(const struct audio_stream __sparse_cache *source,
+					  uint32_t ioffset,
+					  struct audio_stream __sparse_cache *sink,
+					  uint32_t ooffset, uint32_t samples)
+{
+	int32_t *src = source->r_ptr;
+	uint8_t *dst = sink->w_ptr;
+	ae_int32x2 sample = AE_ZERO32();
+	ae_int24x2 sample24 =  AE_ZERO24();
+	uint32_t nmax, i, n, m, left_samples;
+	ae_valign outu = AE_ZALIGN64();
+	ae_valign inu = AE_ZALIGN64();
+
+	ae_int32x2 *in = audio_stream_wrap(source, src + ioffset);
+	ae_int24x2 *out = audio_stream_wrap(sink, dst + ooffset * 3);
+
+	for (left_samples = samples; left_samples; left_samples -= n) {
+		nmax = audio_stream_samples_without_wrap_s32(source, in);
+		n = MIN(left_samples, nmax);
+		nmax = audio_stream_bytes_without_wrap(sink, out) / 3;
+		n = MIN(n, nmax);
+		m = n >> 1;
+		inu = AE_LA64_PP(in);
+		for (i = 0; i < m; i++) {
+			/* load 2 32 bit samples */
+			AE_LA32X2_IP(sample, inu, in);
+			sample24 = AE_MOVINT24X2_FROMF32X2(sample);
+			AE_SA24X2_IP(sample24, outu, out);
+		}
+		AE_SA64POS_FP(outu, out);
+
+		/* process the left 1 sample to avoid memory access overrun */
+		if (n & 0x01) {
+			AE_L32_IP(sample, (ae_int32 *)in, sizeof(ae_int32));
+			sample24 = AE_MOVINT24X2_FROMF32X2(sample);
+			AE_SA24_IP(sample24, outu, out);
+		}
+
+		in = audio_stream_wrap(source, in);
+		out = audio_stream_wrap(sink, out);
+	}
+
+	return samples;
+}
+
+/* 2x24bit samples are packed into 3x16bit samples for hda link dma */
+/* Keep C version now since there are some bugs need to fix, waitting for updating from Rander*/
+static int pcm_convert_s24_c32_to_s24_c24_link_gtw(const struct audio_stream __sparse_cache *source,
+						   uint32_t ioffset,
+						   struct audio_stream __sparse_cache *sink,
+						   uint32_t ooffset, uint32_t samples)
+{
+	int32_t *src = source->r_ptr;
+	uint16_t *dst = sink->w_ptr;
+	int processed;
+	int nmax, i = 0, n = 0;
+
+	src += ioffset;
+	assert(ooffset == 0);
+	for (processed = 0; processed < samples; processed += n) {
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
+		n = samples - processed;
+		nmax = audio_stream_bytes_without_wrap(source, src) >> BYTES_TO_S32_SAMPLES;
+		n = MIN(n, nmax);
+		nmax = audio_stream_bytes_without_wrap(sink, dst) / 3;
+		n = MIN(n, nmax);
+		for (i = 0; i < n; i += 2) {
+			*dst = (*src >> 8) & 0xFFFF;
+			dst++;
+			*dst = (*src & 0xFF << 8) | ((*(src + 1) >> 16) & 0xFF);
+			dst++;
+			*dst = *(src + 1) & 0xFFFF;
+			dst++;
+			src += 2;
+		}
```

Comment:
can it happen that you were an odd number of samples from the wrapping? They you'd have to check for a remaining sample here?

Reply:
Yes, there will be memory overrun risk while processing odd samples  in this way. But as mentioned in the comments, just keep the C version as in pcm_converter_generic.c here to pass compile and make the other 8 functions can be used in Zephyr + ipc4, because there are still some bugs in this function. After discussing with @RanderWang, I will implement it after him fix  those errors.